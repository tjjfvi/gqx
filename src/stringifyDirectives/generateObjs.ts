
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
