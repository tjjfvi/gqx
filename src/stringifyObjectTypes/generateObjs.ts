
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
