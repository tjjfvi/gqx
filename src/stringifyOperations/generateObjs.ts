
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
