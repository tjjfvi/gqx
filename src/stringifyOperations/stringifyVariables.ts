
import { Context, Obj } from "."
import stringifyId from "./stringifyId";

export default (objs: Obj[], ctx: Context) =>
  Object.entries(ctx.config.vars || {}).map(([k, v]) => {
    if(!(v instanceof Array))
      throw new Error("config.vars[string] must be an array");
    let [f, e = true] = v;
    if(typeof f !== "string")
      throw new Error("config.vars[string][0] must be a string");
    if(typeof e !== "boolean")
      throw new Error("config.vars[string][1] must be a boolean");
    if(e) ctx.exports.push(k);
    return `const ${k} = {\n${objs.map(o =>
      `  ${o.prop}: {\n${o.props.map(p =>
        `    ${p.id.prop}: ${v[0]}<typeof ${stringifyId(p.id)}>(${stringifyId(p.id)}),\n`
      ).join("")}  },\n`
    ).join("")}}`
  }).join("\n\n")
