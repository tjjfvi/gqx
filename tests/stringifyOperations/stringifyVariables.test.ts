import generateObjs from "../../src/stringifyOperations/generateObjs"
import stringifyVariables from "../../src/stringifyOperations/stringifyVariables";

test.each([
  ["undefined", {}],
  ["empty", { vars: {} }],
  ["one", { vars: { gqx: ["_gqx"] } }],
  ["two", { vars: { gqx1: ["_gqx1"], gqx2: ["_gqx2"] } }],
])("%s", (_, config) => {
  let [objs, ctx] = generateObjs(utils.getOperations(["extensive"]));
  ctx.config = config;
  let code = stringifyVariables(objs, ctx);
  expect(code).toParse().toMatchSnapshot();
})
