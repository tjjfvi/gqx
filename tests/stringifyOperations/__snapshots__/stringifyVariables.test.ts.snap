// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty 1`] = `""`;

exports[`one 1`] = `
"const gqx = {
  query: {
    getAuthor: _gqx<typeof Query$getAuthor>(Query$getAuthor),
    getBook: _gqx<typeof Query$getBook>(Query$getBook),
    listBooks: _gqx<typeof Query$listBooks>(Query$listBooks),
  },
  mutation: {
    addReview: _gqx<typeof Mutation$addReview>(Mutation$addReview),
  },
}"
`;

exports[`two 1`] = `
"const gqx1 = {
  query: {
    getAuthor: _gqx1<typeof Query$getAuthor>(Query$getAuthor),
    getBook: _gqx1<typeof Query$getBook>(Query$getBook),
    listBooks: _gqx1<typeof Query$listBooks>(Query$listBooks),
  },
  mutation: {
    addReview: _gqx1<typeof Mutation$addReview>(Mutation$addReview),
  },
}

const gqx2 = {
  query: {
    getAuthor: _gqx2<typeof Query$getAuthor>(Query$getAuthor),
    getBook: _gqx2<typeof Query$getBook>(Query$getBook),
    listBooks: _gqx2<typeof Query$listBooks>(Query$listBooks),
  },
  mutation: {
    addReview: _gqx2<typeof Mutation$addReview>(Mutation$addReview),
  },
}"
`;

exports[`undefined 1`] = `""`;
